### PR TITLE
chore: replace current navigation with Voyager [AR-2012]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2012" title="AR-2012" target="_blank">TaskAR-2012</a>  [Spike] Navigation issues -> Fix
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Currently, we have some issues related to the navigation, like: typesafety, passing arguments forth and back, deeplink handling, transitions, overcomplicated code.

### Solutions

This is just a proof of concept so there still may be things that don’t work properly, also tests doesn’t fully pass, because I focused on checking how it would work in our app, not implementing the whole thing as it should be.

Hilt with ViewModel doesn’t support AssistedInject, and that’s one of the main points of using this library (typesafety), on the other hand library offers a substitute in a form of ScreenModel, but unfortunately it bypasses the whole ViewModelComponent and thus we couldn’t keep using ViewModel scope, and it’s used a lot in our app. The chosen solution works with the use of SavedStateHandle, which is the only parameter allowed by Hilt to be passed by AssistedInject-ish way, so there is a new method `getAssistedViewModel` (instead of standard `hiltViewModel` Voyager lib has its own `getViewModel` which works with their `Screen` approach), this new method takes `param` which can be of `Any` type and it’s injected into ViewModel using SavedStateHandle. Passing arguments back is easy - using simply lambdas.
All NavHosts and standard NavGraphControllers are replaced with Navigators, there was no way of combining both in the same app (too many crashes and issues, it was easier to just replace all). Some NavHosts, mainly the ones that shouldn’t be created as a navigation elements in the first place, are replaced by way simpler solutions. For instance, when we have just some composables that should be replaced depending on the state, it’s exactly how Crossfade works, so there is no need to introduce another navigation level.
Handling deep links also is simplified, instead of complex string navigation attributes, there is just a sealed class with current deep link state and everything is handled in the one place.

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

Automated tests are not yet adjusted at this point, there is no sense in spending time before deciding whether we want such a solution at all.

#### How to Test

Just build and open the app and start navigating to different screens to see if everything is correct.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by trOnk12** [comment]
I appreacite all the work you did and I am impressed the level of knowledge you have in such a quick time about the Hilt and the Navigation, I honestly did not read all the code but here is what I think : 

1. Do we really have that much problems with navigation ? Maybe I was the lucky one but I did not experience lots of issues with it. What does this Voyager solves ? 
2. This PR adds lots of magic like AssistedViewModel.kt, that is not so easily understood, thinking ahead perhaps some less experience develops will jump into the AR making for some of them code really hard to understand. It could be solved by adding some documentation to our repo that will guide new developers or anyone starting to work with AR how to navigate and what the current problems are, but it also adds a documentation that needs to be kept up to date and maintaned. 
3. Another question, will it be maintaned by the owner of the voyager ? Jetpack compose is still very fast in progress it is not that stable , changes will come in a fast pace, will the owner of the Voyager maintain it and keep it up to date with Jetpack Compose release ? 
4. Perhaps we can take inspiration from the voyager and write less fat library of our own that would be maintaned by us so that we can keep it up to date with a Jetpack Compose releases.

**Comment by saleniuk** [comment]
> I appreacite all the work you did and I am impressed the level of knowledge you have in such a quick time about the Hilt and the Navigation, I honestly did not read all the code but here is what I think :
> 
> 1. Do we really have that much problems with navigation ? Maybe I was the lucky one but I did not experience lots of issues with it. What does this Voyager solves ?
> 2. This PR adds lots of magic like AssistedViewModel.kt, that is not so easily understood, thinking ahead perhaps some less experience develops will jump into the AR making for some of them code really hard to understand. It could be solved by adding some documentation to our repo that will guide new developers or anyone starting to work with AR how to navigate and what the current problems are, but it also adds a documentation that needs to be kept up to date and maintaned.
> 3. Another question, will it be maintaned by the owner of the voyager ? Jetpack compose is still very fast in progress it is not that stable , changes will come in a fast pace, will the owner of the Voyager maintain it and keep it up to date with Jetpack Compose release ?
> 4. Perhaps we can take inspiration from the voyager and write less fat library of our own that would be maintaned by us so that we can keep it up to date with a Jetpack Compose releases.

1. The main issue that I have is that current navigation is hard to read and understand, it has boilerplate and some workarounds because of the way and limitations they offered with the standard compose navigation. It's becoming hard to keep track of what parameters are in the path and which ones can and should you retrieve from savedStateHandle for instance. Also, our initial approach wasn't good at the first place, because now we have some NavHosts that shouldn't even be added and it only adds to the complexity.

2. It can be documented, for sure, but it's not like you will have to deal with the code inside, you also use the current navigation not seeing and knowing what the code inside looks like. 😄 It's implemented once and you can just use it and to be honest mostly it's just the duplicate of what they did with their ScreenModel but adjusted to use with our ViewModels. 
If you look through it, you will notice that it actually makes it easier for less experienced developers to use navigation. 😉 

3. If the creators stop developing it, then IBM will take it over, that's what the speaker who works for IBM told us on droidcon (he actually got this exact question), so he assured everyone that it will be maintained.

4. This library isn't fat, in fact he also told that they are not afraid to take responsibility for it mainly because it's so small, the core of this library has like only a dozen classes. 😄

**Comment by trOnk12** [comment]
> 

I am convinced :P Could you show some code snippets that you are reffering to the current problems with the navigation ? Because I am not really aware of it and perhaps show how this library will solve it ?

File: app/src/main/kotlin/com/wire/android/ui/authentication/login/LoginScreen.kt [hunk 1]
```diff
@@ -115,11 +130,11 @@ enum class LoginTabItem(@StringRes override val titleResId: Int) : TabItem {
...
-        LoginContent(onBackPressed = { }, ssoLoginResult = null)
-    }
-}
+//@Preview
+//@Composable
+//private fun LoginScreenPreview() {
+//    WireTheme(isPreview = true) {
+//        LoginContent(onBackPressed = { }, ssoLoginResult = null)
+//    }
+//}
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/authentication/login/LoginScreen.kt:139]: commented code

**Review by borichellow — COMMENTED** [review]
Man, that is a massive piece of work and it looks much better than before!!! 💪 
need also run some tests to avoid regression, but cant do it now (app crashes on incoming call every time I open it :) )

**Comment by saleniuk** [comment]
> Could you show some code snippets that you are reffering to the current problems with the navigation ? Because I am not really aware of it and perhaps show how this library will solve it ?

For instance, when you navigate, you have to blindly pass the arguments to `getRouteWithArgs`, then handle them by overriding this method in `NavigationItem` properly and create a string route out of it, which sometimes is like `base/user_id` and sometimes has to be `base?userId=userId` so you need to be very careful. In voyager navigation items have its arguments as constructor parameters so you always know which ones to pass when navigating to a specific place. Passing arguments when navigating back currently is even worse - we have to update the savedStateHandle of NavBackStackEntry and then after creating the ViewModel, update the SavedStateHandle of this ViewModel.

File: app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt [hunk 2]
```diff
@@ -135,8 +137,8 @@ abstract class CreateAccountBaseViewModel(
         viewModelScope.launch {
             navigationManager.navigate(
                 NavigationCommand(
-                    NavigationItem.Login.getRouteWithArgs(),
-                    BackStackMode.CLEAR_TILL_START
+                    VoyagerNavigationItem.Login(),
+                    BackStackMode.REMOVE_CURRENT
```
[the remark this answers is not in the document]
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/authentication/create/common/CreateAccountBaseViewModel.kt:141]: My mistake, I was checkin if we can reduce the amount of BackStackModes, but forgot to bring it back 🙈

**Comment by saleniuk** [comment]
The main question now is: do we want to go this way and dedicate even more time to make it work properly, avoid regressions and adjust existing tests or is it something that we can live without and postpone it until we implement more important stuff?

**Comment by Raynafs** [comment]
Hi Did you manage to implement deep links with voyager?

**Pull request closed without merging**